### PR TITLE
[16.0][MIG/IMP] password_security

### PR DESCRIPTION
**Pull request by astirpe**
Together with the module migration, I'm proposing a revised version of the module:

- [x] Removed old widget, replaced by "Password Meter" widget by standard Odoo
- [x] Unify field `password_length` with standard field `minlength`, to avoid incompatibility with standard module `auth_password_policy_portal`
- [x] Rewrite tests for controller, needed because of new Odoo 16 changes
- [x] Includes https://github.com/OCA/server-auth/pull/492
- [x] Includes https://github.com/OCA/server-auth/pull/494
- [x] Remove password estimate functionality (see https://github.com/OCA/server-auth/pull/482#issuecomment-1653293088)
- [x] Fixes warning in tests https://github.com/OCA/server-auth/issues/141

### CODE REVIEW

**Comment by amh-mw** [comment]
@astirpe Nice job; tremendous amount of work. It feels like it should be six pull requests. If you would be willing to break it up, I'd be willing to review all the parts. I think it's a bit dangerous to do it all in one go, especially with the unit tests being rewritten. How would you feel about fixing each of the items you marked for back-porting in 15.0 as separate pull requests and getting those merged first, so that your migration commit isn't quite so expansive?

**Comment by astirpe** [comment]
@amh-mw, yes that's a good idea!

I backported two of the fixes: https://github.com/OCA/server-auth/pull/491 (constraint of password_estimate range) and https://github.com/OCA/server-auth/pull/492 (Reset Password issue)

The fix for https://github.com/OCA/server-auth/issues/89 is the one already discussed in https://github.com/OCA/server-auth/pull/448 and that's still an open PR.

The fix for https://github.com/OCA/server-auth/issues/141 is just a consequence of the complete rewrite of the mocked tests.

For the rest, I don't think the other points should be backported to V15.
Once the above PRs for V15 will be merged, I will include the related commits in this PR. Is that ok for you?

**Comment by amh-mw** [comment]
> Once the above PRs for V15 will be merged, I will include the related commits in this PR. Is that ok for you?

I think it would make more sense to rebase against 15.0 afterwards, so that you just have the one migration commit.

File: password_security/models/res_company.py
Comment (amh-mw) [inline password_security/models/res_company.py]: Do you need to provide a migration for this field? Maybe something like:

```
ICP = self.env['ir.config_parameter']
ICP.set_param('auth_password_policy.minlength', max(
    ICP.get_param('auth_password_policy.minlength'),
    *self.env['res.company'].search([]).mapped('password_length'))
))
```

File: password_security/controllers/main.py
Comment (amh-mw) [inline password_security/controllers/main.py]: On a big change like this, I would leave style items like this alone to make the diff smaller. Odoo codebase seems to prefer the explicit form anyway.

File: password_security/views/res_config_settings_views.xml
Comment (amh-mw) [inline password_security/views/res_config_settings_views.xml:17]: Seems weird to continue to hide `minlength` now that it has been adopted over `password_length`.

**Review by amh-mw — CHANGES_REQUESTED** [review]
Just poking around a bit while waiting for the maintainers to merge the 15.0 pull requests. I'll give it a more thorough review after.

File: password_security/models/res_company.py
Comment (astirpe) [reply password_security/models/res_company.py]: Yes, seems a good idea :+1:  Thanks!

File: password_security/views/res_config_settings_views.xml
Comment (astirpe) [reply password_security/views/res_config_settings_views.xml:17]: It's just because I wanted to embed the `minlength` inside the group of fields "Password Policy"

https://github.com/OCA/server-auth/blob/6eab9a5d01c5fa49263ea9aba33bade290bfc9cf/password_security/views/res_config_settings_views.xml#L53-L57

This way it looks&feels better, otherwise the `minlength` will stay detached from the rest of the fields.

File: password_security/models/res_company.py
Comment (astirpe) [reply password_security/models/res_company.py]: Implemented. Migration tests could be executed locally by adding `password_length = fields.Integer(default=12)` to res.company

**Comment by astirpe** [comment]
No idea why OCB test now fails. Maybe a temporary glitch?

**Comment by Abranes** [comment]
@astirpe Could you rebase from 16.0 to resolve the conflicts?

**Comment by astirpe** [comment]
Still test in OCB failing, I cannot even reproduce the failure with OCB locally.

**Comment by astirpe** [comment]
Rebased. And this is another surprise: now it's Odoo failing, not OCB anymore

File: password_security/models/res_users.py
Comment (kmagusiak) [inline password_security/models/res_users.py]: This method can be called as a (portal) user from `_check_credentials` with a replacement hash. In that case, the user does not have permissions to write.

```suggestion
        self.sudo().write({"password_history_ids": [(0, 0, {"password_crypt": pw})]})
```

Comment (astirpe) [reply password_security/models/res_users.py]: @kmagusiak thank you!
I made a fix in commit https://github.com/OCA/server-auth/pull/482/commits/4b1b2686d841b73f2594103d8218a17c989fa0c5 by setting the proper rights for portal user. Is that working for you?

Comment (kmagusiak) [reply password_security/models/res_users.py]: I haven't tested, but it looks like it should work.

**Review by jaenbox — CHANGES_REQUESTED** [review]
I have tested the password_security module from the runboat. When enabling the module and enabling MFA, I encounter an incorrect behavior. MFA configuration is set up correctly. The problem arises when attempting a second login with this user. A 500 error occurs. In the log, it shows 'Login successful for xxxx,' but later, a singleton error is displayed: 'ValueError: Expected singleton: res.users()'.

**Comment by astirpe** [comment]
@jaenbox thank you! I just reproduced the error and I'll fix it soon:

```
Traceback (most recent call last):
  File "/opt/odoo/odoo/models.py", line 5129, in ensure_one
    _id, = self._ids
ValueError: not enough values to unpack (expected 1, got 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/odoo/odoo/http.py", line 1584, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/opt/odoo/odoo/service/model.py", line 133, in retrying
    result = func()
  File "/opt/odoo/odoo/http.py", line 1611, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/opt/odoo/odoo/http.py", line 1725, in dispatch
    return self.request.registry['ir.http']._dispatch(endpoint)
  File "/opt/odoo/addons/website/models/ir_http.py", line 235, in _dispatch
    response = super()._dispatch(endpoint)
  File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 154, in _dispatch
    result = endpoint(**request.params)
  File "/opt/odoo/odoo/http.py", line 697, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/opt/odoo/addons/website/controllers/main.py", line 186, in web_login
    return super().web_login(*args, **kw)
  File "/opt/odoo/odoo/http.py", line 697, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/mnt/data/odoo-addons-dir/password_security/controllers/main.py", line 31, in web_login
    if not request.env.user._password_has_expired():
  File "/mnt/data/odoo-addons-dir/password_security/models/res_users.py", line 144, in _password_has_expired
    self.ensure_one()
  File "/opt/odoo/odoo/models.py", line 5132, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: res.users()
```

**Comment by astirpe** [comment]
@jaenbox done, would you give it a try?

**Comment by jaenbox** [comment]
> @jaenbox done, would you give it a try?

@astirpe It's works!!!

**Comment by astirpe** [comment]
@amh-mw @jaenbox @kmagusiak since you already reviewed and tested this old PR, do you mind to approve it? Thanks in advance!

**Review by kmagusiak — APPROVED** [review]

**Review by amh-mw — APPROVED** [review]

**Comment by zuher83** [comment]
Any news? 🤔

**Comment by thomaspaulb** [comment]
[test was failing](https://github.com/OCA/server-auth/actions/runs/5436772234/jobs/9886859848#step:8:361)

/ocabot rebase

**Comment by thomaspaulb** [comment]
@astirpe Could you check out the failing test, not sure why it appeared

File: password_security/views/res_config_settings_views.xml
Comment (hwangh95) [reply password_security/views/res_config_settings_views.xml:17]: How about instead of removing the entire box that contains the label, we reuse it? There might have been some customization within the settings box that gets removed otherwise.

Something like this?
```
<!-- Remove label for uniform styling -->
<xpath expr="//label[@for='minlength']" position="replace"/>


<!-- Add an id for easier lookup -->
<xpath expr="//field[@name='minlength']/ancestor::div[hasclass('o_setting_box')]" position="attributes">
    <attribute name="id">password_policy</attribute>
</xpath>

<!-- Move the settings box to desired location -->
<xpath expr="//div[@id='enable_password_reset']" position="after">
    <xpath expr="//div[@id='password_policy']" position="move"/>
</xpath>


<!-- Ensure our settings will come before any former customization -->
<xpath expr="//div[@id='password_policy']//div[hasclass('o_setting_right_pane')]/*" position="before">
    <!-- Add all the settings here -->
    <span id="minlength">
         Minimum number of characters
    </span>
</xpath>


<!-- Add oe_inline attribute to minlength -->
<xpath expr="//field[@name='minlength']" position="attributes">
  <attribute name="class">oe_inline</attribute>
</xpath>

<!-- Move the minlength field to desired location -->
<xpath expr="//span[@id='minlength']" position="inside">
    <xpath expr="//field[@name='minlength']" position="move"/>
</xpath>
```

It might seem like a lot of work, but I personally feel like we should try to minimize making nodes disappear

File: password_security/tests/test_change_password.py
Comment (hwangh95) [inline password_security/tests/test_change_password.py:100]: Odoo stores datetime fields with time granularity of seconds. The tests are failing because the _change_password is processed fast enough for some of the history to be within the same second.

You need to add either sleeps or manually modify the 'date' field in the password_history_ids to be distinct, otherwise it will be ordered by id asc. You could also add id desc as a sort of fallback to the _order.

**Review by hwangh95 — CHANGES_REQUESTED** [review]
Please change either the tests or the _order of 'res.users.pass.history' to have the tests pass.

**Comment by cmarrero** [comment]
The reuse of previous passwords fails if user try with their last password, after the message that validate you are using an old password if you enter a new one gives you a invalid token error and you have to send again another reset to make it works @astirpe

File: password_security/tests/test_signup.py
Comment (hwangh95) [inline password_security/tests/test_signup.py:153]: Because of auth_signup_verify_email, the error message will be "That does not seem to be an email address" rather than "Signup: no name or partner given for new user". I would try with a "valid" login format like "test@test.com"

**Review by hwangh95 — CHANGES_REQUESTED** [review]
Tests are failing because of auth_signup_verify_email

File: password_security/tests/test_signup.py
Comment (hwangh95) [reply password_security/tests/test_signup.py:153]: Gaaaah auth_signup_verify_email just eats the error and throws a generic one. Maybe you could do some sort of check against the installed module? Or make a small PR to auth_signup_verify_email to add the thrown error to the error context.

File: password_security/views/res_config_settings_views.xml
Comment (astirpe) [reply password_security/views/res_config_settings_views.xml:17]: Yes thanks! Done in https://github.com/OCA/server-auth/pull/482/commits/b72dde148032d39043c5fd8b2bbe31461821da3f, would you check?

Comment (hwangh95) [reply password_security/views/res_config_settings_views.xml:17]: Looks mostly good to me! There's the comment that still says 

I don't know if you wanted the password_security settings to come before or after the previous nodes.

File: password_security/models/res_company.py
Comment (hwangh95) [inline password_security/models/res_company.py:13]:
```suggestion
```
I don't know what the solution is unfortunately. I tried setting all the defaults to zero and setting the company values to the "correct" defaults on each of the tests but then the main company will have default 0 values.

I thought maybe we could use ir.defaults to put the defaults back, but the default values will only be set on a new company creation.

We can't use a post_install_hook since test suites are run with a post-install tag...

Do you have any suggestions?

**Review by hwangh95 — CHANGES_REQUESTED** [review]
I've been doing some testing on an odoo.sh instance and unfortunately with the defaults set on company fields, anyone who installs this module on something with a stricter test suite will fail the base tests.

File: password_security/models/res_company.py
Comment (astirpe) [reply password_security/models/res_company.py:13]: I would go for setting `password_expiration` default to 0 and update the readme and the tests accordingly. What do you think?

Do you know if `password_expiration` is the only field to be fixed?

File: password_security/tests/test_signup.py
Comment (hwangh95) [inline password_security/tests/test_signup.py:142]:
```suggestion
                    "login": "test@test.com",
                    "password": "test123",
```
Looking at auth_signup_verify_email, it seems like it only redirects the signup for passwordless signups. So If you set a password, it should be handled by web_auth_signup and your exception handler.

File: password_security/tests/test_signup.py
Comment (hwangh95) [inline password_security/tests/test_signup.py]:
```suggestion
        self.assertIn(
            "Signup: no name or partner given for new user",
            response.text,
        )
```
By adding the "password" param, we can go back to the previous assertion

Comment (hwangh95) [inline password_security/tests/test_signup.py:155]: To cover the code in [password_security/controllers/main.py#L57-L61](https://github.com/OCA/server-auth/pull/482/files#diff-2c1c11f2ce9b2571602f9b0185b57a72e738dd34a5381a6478d1f27a2264ee87R57-R60), you can add some tests like these
```suggestion
        )
        self.assertIn("X-Frame-Options", response.headers)
        self.assertEqual(
            response.headers["X-Frame-Options"],
            "SAMEORIGIN"
        )
        
        self.assertIn("Content-Security-Policy", response.headers)
        self.assertEqual(
            response.headers["Content-Security-Policy"],
            "frame-ancestors 'self'"
        )
```

**Review by hwangh95 — CHANGES_REQUESTED** [review]
Found some ways to get around the auth_signup_verify_email issue

File: password_security/models/res_company.py
Comment (hwangh95) [reply password_security/models/res_company.py:13]: Unfortunately it's all of them 🥲
base test errors
check password error

It's because the base test suite passwords don't fit the password requirements...

Comment (amh-mw) [reply password_security/models/res_company.py:13]: See https://github.com/odoo/odoo/pull/114858 for some commentary from Odoo developers on this problem.

File: password_security/tests/test_signup.py
Comment (astirpe) [reply password_security/tests/test_signup.py:142]: You are right, thank you!

File: password_security/models/res_company.py
Comment (astirpe) [reply password_security/models/res_company.py:13]: Regarding this problem, for the time being, I would revert https://github.com/OCA/server-auth/pull/482/commits/58adcf51fd527c1621b77b2d60842e89dd6d7ad9 and I would avoid to solve the problem within this PR. Later on we could open a separate PR in case https://github.com/odoo/odoo/pull/114858 will not go any further.
Is that ok with you all?

**Comment by cmarrero** [comment]
After upgrade and enter Settings gives the following error with the refactor:

RPC_ERROR
Odoo Server Error
Traceback (most recent call last):
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/http.py", line 1584, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/service/model.py", line 133, in retrying
    result = func()
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/http.py", line 1611, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/http.py", line 1815, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/Users/cmarrero/PycharmProjects/odoo/addons/website/models/ir_http.py", line 235, in _dispatch
    response = super()._dispatch(endpoint)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/addons/base/models/ir_http.py", line 154, in _dispatch
    result = endpoint(**request.params)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/http.py", line 697, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/Users/cmarrero/PycharmProjects/odoo/addons/web/controllers/dataset.py", line 42, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/Users/cmarrero/PycharmProjects/odoo/addons/web/controllers/dataset.py", line 33, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/api.py", line 457, in call_kw
    result = _call_kw_model(method, model, args, kwargs)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/api.py", line 430, in _call_kw_model
    result = method(recs, *args, **kwargs)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/addons/base/models/ir_ui_view.py", line 2491, in get_views
    result['views'] = {
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/addons/base/models/ir_ui_view.py", line 2492, in <dictcomp>
    v_type: self.get_view(
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/addons/base/models/ir_ui_view.py", line 2676, in get_view
    result = dict(self._get_view_cache(view_id, view_type, **options))
  File "<decorator-gen-48>", line 2, in _get_view_cache
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/tools/cache.py", line 90, in lookup
    value = d[key] = self.method(*args, **kwargs)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/addons/base/models/ir_ui_view.py", line 2636, in _get_view_cache
    arch, view = self._get_view(view_id, view_type, **options)
  File "/Users/cmarrero/PycharmProjects/odoo/addons/web/models/models.py", line 245, in _get_view
    arch, view = super()._get_view(view_id, view_type, **options)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/addons/base/models/ir_ui_view.py", line 2583, in _get_view
    arch = view._get_combined_arch()
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/addons/base/models/ir_ui_view.py", line 999, in _get_combined_arch
    arch = root.with_prefetch(tree_views._prefetch_ids)._combine(hierarchy)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/addons/base/models/ir_ui_view.py", line 940, in _combine
    combined_arch = view.apply_inheritance_specs(combined_arch, arch)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/addons/base/models/ir_ui_view.py", line 877, in apply_inheritance_specs
    self._raise_view_error(str(e), specs_tree)
  File "/Users/cmarrero/PycharmProjects/odoo/odoo/addons/base/models/ir_ui_view.py", line 754, in _raise_view_error
    raise err from from_exception
ValueError: Element '<xpath expr="//span[@id=&#39;minlength&#39;]">' cannot be located in parent view

The above server error caused the following client error:
RPC_ERROR: Odoo Server Error
    at makeErrorFromResponse (http://localhost:8069/web/assets/10737-0b43f28/web.assets_backend.min.js:985:163)
    at XMLHttpRequest.<anonymous> (http://localhost:8069/web/assets/10737-0b43f28/web.assets_backend.min.js:993:13)

**Comment by astirpe** [comment]
@cmarrero if you are getting the error in a production server I would suggest you to restore the old code (before the config settings refactoring).

Instead if you are on a test server you can try this:

- Login as admin with enabled developer mode
- Go to Settings > Technical > User interface > Views
- Search view named "res.config.settings.form.password_security" and delete it
- Go to Apps
- Search module Password Security and click on Upgrade button
- Verify that the view named "res.config.settings.form.password_security" is re-created
- Check whether the error does not occur again

File: password_security/models/res_company.py
Comment (hwangh95) [reply password_security/models/res_company.py:13]: Makes sense. I think I'll end up having to make a custom module to prevent test failures on my end for now.

**Comment by astirpe** [comment]
I just found out that when a new user is trying to signup by entering a very large password, there is a critical performance issue.
It's caused by the adoption of `zxcvbn` library, in particular the execution of this line https://github.com/dwolfhub/zxcvbn-python/blob/master/zxcvbn/__init__.py#L25 is very slow. That could potentially cause a DDOS attack in my opinion.

Probably that's also the reason of https://github.com/OCA/server-auth/issues/487 but I didn't investigate in deep that issue.

So I'm going to remove the password estimate check functionality from this module and also remove the dependency on `zxcvbn`, because the password estimation is already provided by "Password Meter" widget by standard Odoo.

I hope is ok with you all.

**Comment by hwangh95** [comment]
Are there any updates on this PR? Or is it good to go?

**Comment by astirpe** [comment]
Would be good for me if this PR gets merged soon

File: password_security/views/res_config_settings_views.xml
Comment (hwangh95) [inline password_security/views/res_config_settings_views.xml]: I was doing some testing on my own instance and noticed this portion was slightly off
```suggestion
                      <div class="mt16">
                          <span id="minlength">
                               Minimum number of characters
                          </span>
                      </div>
                  </div>
```

**Review by hwangh95 — CHANGES_REQUESTED** [review]
Small UI change

File: password_security/models/res_company.py
Comment (hwangh95) [reply password_security/models/res_company.py:13]: @astirpe How do you feel about having an `enable_password_security` field on res.config.settings that is related to company_id?

Then we can enable it for tests and have the setting_box attr be invisible depending on it

Comment (astirpe) [reply password_security/models/res_company.py:13]: @hwangh95 seems a good idea to me, however I'm very busy on a project these days. Is there any possibility for you to take over this PR for continuing the development? Would be really appreciated and I could review it later on.

**Comment by zuher83** [comment]
Ready to migrate ?

**Comment by antonag32** [comment]
Hello, is there anything left to do before merging? I could help with any tasks left.

**Pull request merged**
